sdks/python/apache_beam/runners/portability/local_job_service.py

```diff
@@ -185,55 +152,89 @@ class BeamJob(threading.Thread):
     The current state of the pipeline is available as self.state.
     """
 
-  def __init__(self,
-               job_id,
-               pipeline_options,
-               pipeline_proto):
+  def __init__(self, pipeline_proto):
     super(BeamJob, self).__init__()
-    self._job_id = job_id
-    self._pipeline_options = pipeline_options
     self._pipeline_proto = pipeline_proto
     self._state = None
-    self._state_change_callbacks = []
-    self._last_log_message = None
-    self._log_callbacks = [lambda msg: setattr(self, '_last_log_message', msg)]
+    self._logs = []
```

Comment:
I still don't think that we should be storing all logs. What if we allowed connecting *before* the job is started to ensure all logs were seen. Otherwise, one gets the logs from the point at which one connected (which better matches the intention). 

Similarly for states, as long as one gets a response for the current state, there's no need to replay all past states. In this case we would go back to the callback logic, but maybe remove the last_log_message which is the fishy one, and fix the runner to start listening to logs before it starts the job (which we'll want to do for other jobs servers as well).

Reply:
Gotcha, I was under the impression that this was a testing only class. I understand how troublesome long running jobs can be on memory usage. I changed it the semantics back to the queue usage and moved the Get(State|Message)Stream to before the job is run.